src/execution/operator/projection/physical_tableinout_function.cpp

```diff
@@ -52,15 +56,152 @@ unique_ptr<GlobalOperatorState> PhysicalTableInOutFunction::GetGlobalOperatorSta
+	D_ASSERT(mapping_data.validity.AllValid());
+	auto mapping_array = (sel_t *)mapping_data.data;
+
+	// We can directly use this column as a selection vector
+	SelectionVector sel_vec(mapping_array);
+
+	// Add the projected columns, and apply the selection vector
+	for (idx_t project_idx = 0; project_idx < projected_input.size(); project_idx++) {
+		auto source_idx = projected_input[project_idx];
+		D_ASSERT(source_idx < input.data.size());
+		auto target_idx = map_idx + project_idx;
+
+		auto &target_column = out.data[target_idx];
+		auto &source_column = input.data[source_idx];
+
+		target_column.Slice(source_column, sel_vec, intermediate.size());
+		// Since our selection vector is using temporary allocated data, we need to
+		// immediately flatten this column, so we don't run the risk of the dictionary vector
+		// outliving the selection vector
+		target_column.Flatten(intermediate.size());
+	}
+}
+
+OperatorResultType PhysicalTableInOutFunction::ExecuteWithMapping(ExecutionContext &context, DataChunk &input,
+                                                                  DataChunk &chunk, TableInOutLocalState &state,
+                                                                  TableFunctionInput &data) const {
+	// Create a duplicate of 'chunk' that contains one extra column
+	// this column is used to register the relation between input tuple -> output tuple(s)
+	const auto base_columns = chunk.ColumnCount() - projected_input.size();
+
+	DataChunk intermediate_chunk;
+	// Create an empty DataChunk that has room for the input + the mapping vector
+	auto chunk_types = chunk.GetTypes();
+	vector<LogicalType> intermediate_types;
+	intermediate_types.reserve(base_columns + 1);
+	intermediate_types.insert(intermediate_types.end(), chunk_types.begin(), chunk_types.begin() + base_columns);
+	intermediate_types.emplace_back(LogicalType::UINTEGER);
+	// We initialize this as empty
+	intermediate_chunk.InitializeEmpty(intermediate_types);
+	// And only allocate for our mapping vector
+	intermediate_chunk.data[base_columns].Initialize();
+
+	// Initialize our output chunk
+	for (idx_t i = 0; i < base_columns; i++) {
+		intermediate_chunk.data[i].Reference(chunk.data[i]);
+	}
+	intermediate_chunk.SetCardinality(chunk.size());
+
+	// Let the function know that we expect it to write an in-out mapping for rowids
+	data.add_in_out_mapping = true;
+	auto result = function.in_out_function(context, data, input, intermediate_chunk);
+	chunk.SetCardinality(intermediate_chunk.size());
+
+	// Move the result into the output chunk
+	for (idx_t i = 0; i < base_columns; i++) {
+		chunk.data[i].Reference(intermediate_chunk.data[i]);
+	}
+
+	auto &mapping_column = intermediate_chunk.data[base_columns];
+	switch (mapping_column.GetVectorType()) {
+	case VectorType::FLAT_VECTOR: {
+		// This is the original vector we created
+		AddProjectedColumnsFromFlatMapping(base_columns, input, intermediate_chunk, chunk);
```

Comment:
There doesn't seem to be much of a performance difference between the flat and generic cases, can we perhaps remove the flat case to simplify the code?